x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/user/GetUserPrivilegesResponse.java

```diff
@@ -82,13 +94,28 @@ public Set<String> getRunAs() {
         return runAs;
     }
 
+    public boolean hasRemoteIndicesPrivileges() {
+        return false == remoteIndex.isEmpty();
+    }
+
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeCollection(cluster, StreamOutput::writeString);
         out.writeCollection(configurableClusterPrivileges, ConfigurableClusterPrivileges.WRITER);
         out.writeCollection(index);
         out.writeCollection(application);
         out.writeCollection(runAs, StreamOutput::writeString);
+        if (out.getVersion().onOrAfter(RoleDescriptor.VERSION_REMOTE_INDICES)) {
+            out.writeCollection(remoteIndex);
+        } else if (hasRemoteIndicesPrivileges()) {
+            throw new IllegalArgumentException(
```

Comment:
Many responses of security APIs, including this one, do not travel across wire. So in practice how this method behaves has no impact. 

It should not be a problem for a mixed cluster because:
1. We do _not_ allow roles with remote indices to be created in a mixed cluster that has node older than 8.6
2. Reserved roles are node local. If a reserved role is updated with remote indices in v8.6. A old node, e.g. 8.5, will keep returning its old definition for this role. While a 8.6 node will return the new definition.

For point 2, it does mean a user with the said reserved role will get different privileges on nodes of different versions. But this is not new. Any changes to reserved roles today can have this effect.

Reply:
Good q! As Yang points out, this won't go on the wire since the action will be executed directly on the coordinating node (I'll send you a Slack thread for more context around local-only security actions) so the reserved roles will be node-local.

> We do not allow roles with remote indices to be created in a mixed cluster that has node older than 8.6

We enforce this [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativeRolesStore.java#L224-L233).

Reply:
Thank you for the link. I was wondering how we can implement that a min version. Maybe it should be in a helper method, but not in scope for this PR.

Reply:
Thanks! What I was missing in my initial comment was that reserved roles are per-node - I was up to date on Point 1, but that's enforce in the Put Role action, which internal processes may not go through. I'm glad you're one step ahead of me on this; I'm just trying to be extra careful because streaming problems are one of the relatively few ways to dig a hole that's really hard to get out of if/when they come up as SDHs.